Sources/web3swift/Transaction/AbstractEnvelope.swift

```diff
@@ -0,0 +1,144 @@
+    /// The type of transaction this envelope represents
+    var type: TransactionType { get }
+
+    // common parameters for any transaction
+    /// the nonce value for the transaction
+    var nonce: BigUInt { get set }
+    /// Blockchain ChainID that this transaction is or will be, signed for.
+    /// Remains optional to support legacy transactions.
+    var chainID: BigUInt? { get set }
+    /// On chain address that this transaction is being sent to
+    var to: EthereumAddress { get set }
+    /// The native value of the transaction in Wei
+    var value: BigUInt { get set }
+    /// Any encoded data accompanying the transaction
+    var data: Data { get set }
+    // Signature data should not be set directly
+    /// signature V compoonent
+    var v: BigUInt { get set }
+    /// signature R compoonent
+    var r: BigUInt { get set }
+    /// signature S compoonent
+    var s: BigUInt { get set }
+
+    // required initializers
+    // for Decodable support
+    /// initializer for creating an EthereumTransaction with the Decodable protocol
+    /// will return an new EthereumTransaction object on success
+    /// thows a Web3.dataError if an error occurs whiel trying to decode a value
+    /// returns nil if a required field is not found in the decoder stream
+    init?(from decoder: Decoder) throws // Decodable Protocol
+
+    // initializes from a raw stream of bytes
+    // can fail if input stream is not of the right size/cannot be decoded
+    /// initializer for creating an EthereumTransaction with raw bytestream data
+    /// will return an new EthereumTransaction object on success
+    /// returns nil if a required field is not found in the decoder stream, or can't be decoded
+    init?(rawValue: Data) // Decode from Ethereum Data
+
+    // pseudo memberwise initializer
+    // accepts all common parameters (full description with default implementation below)
+    init(to: EthereumAddress, nonce: BigUInt?, chainID: BigUInt?, value: BigUInt?,
+         data: Data, v: BigUInt, r: BigUInt, s: BigUInt, options: TransactionOptions?)
+
+    /// Applies the passed options to the transaction envelope
+    ///   - Parameters:
+    ///     - {default}: TransactionOptions struct
+    mutating func applyOptions(_ options: TransactionOptions)
+    ///  Create a TransactionOptions object representing this transaction
+    /// - Returns: A TransactionOptions object with all parameters from the transaction set
+    func getOptions() -> TransactionOptions
+
+    /// Transaction encoder for transmission or signing
+    ///  - Parameters:
+    ///    - {default}: EncodeType enum
+    ///     - when type is .signature the transaction is encoded for signing
+    ///     - when type is .transaction the thransaction is encoded for hashing or transmission to the blockchain
+    /// - Returns: a raw encoding stream representing the transaction, encoded according to it's type
+    func encodeFor(_ type: EncodeType) -> Data?
```

Comment:
changed here. That was my original intended name but went with `encodeFor` to avoid confusion with the existing `encode(forSignature:)` as I was refactoring on the first pass, with the intent of coming back later to change it, and forgot.